pangeo_forge_recipes/recipes/reference_hdf_zarr.py

```diff
@@ -0,0 +1,202 @@
+    xarray, fsspec, zarr, h5py and ujson in your recipe's requirements.
+
+    This class will also produce an Intake catalog stub in YAML
+    format, with the same name as the output json + ".yaml". You
+    can use intake (and intake-xarray) to load the dataset, and this
+    is the recommended way to distribute access.
+
+    SingleHdf5ToZarr parameters:
+
+    :param netcdf_url: location of the original data files. Can be a
+        string (will be glob-expanded) or explicit list of paths or
+        list of glob-strings. May eventually support FilePatterns.
+        Should include protocol specifier, as usually formulated for
+        fsspec.
+    :param netcdf_storage_options: dict of kwargs for creating fsspec
+        instance to read original data files
+    :param inline_threshold: blocks with fewer bytes than this will be
+        inlined into the output reference file
+
+    MultiZarrToZarr parameters (if accessing more than one HDF):
+
+    :param xarray_open_kwargs: kwargs passed to xarray.open_dataset
+    :param xarray_concat_args: kwargs passed to xarray.concat
+
+    Output parameters:
+
+    :param output_url: where the final reference file is written.
+    :param output_storage_options: dict of kwargs for creating fsspec
+        instance when writing final output
+    :param template_count: the number of occurrences of a URL before it
+        gets made a template. ``None`` to disable templating
+    """
+
+    # TODO: support chunked ("tree") aggregation: would entail processing each file
+    #  in one stage, running a set of merges in a second step and doing
+    #  a master merge in finalise. This would maybe map to iter_chunk,
+    #  store_chunk, finalize_target. The strategy was used for NWM's 370k files.
+
+    # TODO: as written, this rcipe is specific to HDF5 files. fsspec-reference-maker
+    #  also supports TIFF and grib2 (and more coming)
+    netcdf_url: Union[str, List[str]]
+    output_url: str
+    _work_dir: str = ""
+    """temporary store for JSONs"""
+    netcdf_storage_options: dict = field(default_factory=dict)
+    inline_threshold: int = 500
+    output_storage_options: dict = field(default_factory=dict)
+    template_count: Optional[int] = 20
+    xarray_open_kwargs: dict = field(default_factory=dict)
+    xarray_concat_args: dict = field(default_factory=dict)
+
+    def iter_inputs(self) -> Iterable[Hashable]:
+        return ()
+
+    @property
+    def cache_input(self) -> Callable[[Hashable], None]:
+        return no_op
+
+    @property
+    def prepare_target(self) -> Callable[[], None]:
+        """Prepare the recipe for execution by initializing the target.
+        Attribute that returns a callable function.
+        """
+        self._work_dir = tempfile.mkdtemp()
+        return no_op
+
+    def iter_chunks(self) -> Iterable[Hashable]:
+        """Iterate over all target chunks."""
+        return fsspec.open_files(self.netcdf_url, **self.netcdf_storage_options)
```

Comment:
What does this return? Open file objects?

Would nice to plug in FilePattern here.

Reply:
It returns an `OpenFiles` instance, which will open all contained files when encountering a `with`, or can be iterated to expose individual `OpenFile` instances.
What would we do with a FilePattern? We don't have logic to infer coordinates or add them into the chunking scheme. It would be nice in the long run, but tricky.

Reply:
> What would we do with a FilePattern?

Mostly the point is to align with the rest of Pangeo Forge.

I would recommend using this instead

https://github.com/pangeo-forge/pangeo-forge-recipes/blob/c4867719c4866ec46436b5d53d95ceb470a6f72e/pangeo_forge_recipes/recipes/xarray_zarr.py#L882-L884

and then using the input key to look up the url from the FilePattern object, e.g. 

https://github.com/pangeo-forge/pangeo-forge-recipes/blob/c4867719c4866ec46436b5d53d95ceb470a6f72e/pangeo_forge_recipes/recipes/xarray_zarr.py#L167